### PR TITLE
#2521 Navigation tidy up

### LINKED ISSUE #2521: New navigation tidy up
## Is your feature request related to a problem? Please describe.

Each of the routes defined in `mSupplyMobilePage` could easily by mapped to tidy things up!

## Describe the solution you'd like

As above

## Implementation

As above

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
#### BRANCHED FROM #2519 

Fixes #2521 

## Change summary

- Tidies up `mSupplyMobileApp` and the mass amounts of page components!

## Testing

N/A

### Related areas to think about

N/A

### CODE REVIEW

File: src/mSupplyMobileApp.js [hunk 1]
```diff
@@ -1,68 +1,42 @@
...
+import { selectIsSyncing } from './selectors/sync';
+
+import { syncCompleteTransaction, setSyncError, openSyncModal } from './actions/SyncActions';
+import { FinaliseActions } from './actions/FinaliseActions';
 import { UserActions } from './actions';
+import { SupplierCreditActions } from './actions/SupplierCreditActions';
 
+import { Spinner } from './widgets';
+import { ModalContainer, FinaliseModal, LoginModal } from './widgets/modals';
+import { FirstUsePage } from './pages';
```
Comment (joshxg) [inline src/mSupplyMobileApp.js]: just reordered imports ..

File: src/mSupplyMobileApp.js [hunk 2]
```diff
@@ -214,74 +178,8 @@ class MSupplyMobileAppContainer extends React.Component {
...
-            <MainStackNavigator.Screen name={ROUTES.DISPENSARY} component={DispensingPage} />
-            <MainStackNavigator.Screen name={ROUTES.PRESCRIPTION} component={PrescriptionPage} />
-
-            <MainStackNavigator.Screen name={ROUTES.CASH_REGISTER} component={CashRegisterPage} />
-
-            <MainStackNavigator.Screen name={ROUTES.REALM_EXPLORER} component={RealmExplorer} />
-            <MainStackNavigator.Screen name={ROUTES.SETTINGS} component={SettingsPage} />
-            <MainStackNavigator.Screen name={ROUTES.DASHBOARD} component={DashboardPage} />
+          <MainStackNavigator.Navigator initialRouteName={ROUTES.MENU}>
+            {Pages}
```
Comment (joshxg) [inline src/mSupplyMobileApp.js]: `Pages` now an array of page components from `Navigator`

File: src/navigation/Navigator.js [hunk 3]
```diff
@@ -2,7 +2,81 @@
...
+
+export const DEFAULT_SCREEN_OPTIONS = {
+  headerLeft: () => <BackButton />,
+  headerTitleAlign: 'center',
+  headerTitle: MsupplyMan,
+  headerRight: () => <SyncState />,
+  headerStyle: navigationStyles.headerStyle,
+  headerLeftContainerStyle: navigationStyles.headerLeftContainerStyle,
+  headerRightContainerStyle: navigationStyles.headerRightContainerStyle,
+};
```
Comment (joshxg) [inline src/navigation/Navigator.js]: these are the default options for a screen - show the back button, msupply man and sync state.

File: src/navigation/Navigator.js [hunk 4]
```diff
@@ -2,7 +2,81 @@
...
+
+const FINALISABLE_SCREEN_OPTIONS = {
+  headerRight: () => <FinaliseButton />,
```
Comment (joshxg) [inline src/navigation/Navigator.js]: for a finalisable page, show the finalise button instead

File: src/navigation/Navigator.js [hunk 5]
```diff
@@ -2,7 +2,81 @@
...
+  [ROUTES.STOCKTAKE_MANAGER]: StocktakeManagePage,
+  [ROUTES.STOCKTAKE_EDITOR]: StocktakeEditPage,
+  [ROUTES.DISPENSARY]: DispensingPage,
+  [ROUTES.PRESCRIPTION]: PrescriptionPage,
+  [ROUTES.CASH_REGISTER]: CashRegisterPage,
+  [ROUTES.SETTINGS]: SettingsPage,
+  [ROUTES.DASHBOARD]: DashboardPage,
+};
+
+export const Pages = Object.values(ROUTES).reduce((acc, route) => {
```
Comment (joshxg) [inline src/navigation/Navigator.js]: mapping each page to an array. There are some routes without a real page.. they need to be removed

**Comment by wlthomson** [comment]
@joshxg Could we base these on a feature branch? I think at this stage it would be easier to review in one go.

**Comment by joshxg** [comment]
@wlthomson the sub tree is `feature/navigation-update` -> #2518 -> #2520 -> #2522 

- (I think this'll be something we disagree on!)
- I always point the base to the most recent branch as it's a smaller diff 
- it seems most other people prefer this. 
- If it's easier for you to review we can change the base of this branch to `feature/navigation-update`? Do Feel free to do it when you go to review
- Actioning changes is harder and slower - rather than waiting for a smaller review to action changes, have to wait for a large review to action any changes. Then, the revisit takes longer to verify changes made over a large single PR
 - if you review the first(#2518) in order, then it really makes no difference that these are all branched off each other

Again, happy for you to change the base of this to `feature/navigation` and just review this if that's better for you

**Comment by wlthomson** [comment]
@joshxg Hm, I find the main problem with recursive branching is that it can be a bit inefficient just due to the fact that often things brought up in reviews have already been addressed in future commits (and it's difficult to mitigate this by working backwards, as often you don't have the full picture of the changes). 

Re. actioning changes, I feel it's actually a bit of a false economy and more difficult this way as changes to older branches have to be potentially resolved in newer branches (or actioned in the latest branch, which makes things equally hard to follow). 

Moving forward, a good compromise might be to base each change on the `feature` branch, but also add a link containing the diff between this branch and the previous branch. That way reviewers can get a "big picture" view, but also be able to limit their comments to things relevant to this branch only (another compromise would be to do it the other way, and link to a diff between this and the `feature` branch)?

**Comment by joshxg** [comment]
@wlthomson yup it's very inefficient. It's also impossible to predict what every reviewer would like in order to help them to review (i.e. I thought comments might be helpful, you prefer a different diff etc). As I said - no one has given any similar feedback.

- The diff between this branch and the feature branch: https://github.com/openmsupply/mobile/compare/feature/react-navigation-upgrade...%232521-navigation-tidy-up
- You can just head to the "Code" tab and find the branch there, then click "compare" and choose the feature branch to get that one! :trollface: 

I think the root problem here is time-to-review and lack of communication. Shouldn't need to branch if theres no branch!

File: src/pages/CashRegisterPage.js [hunk 6]
```diff
@@ -165,12 +165,12 @@ export const CashRegister = ({
...
+const mapDispatchToProps = dispatch => {
   const onUpdatePaymentType = paymentType =>
     dispatch(updatePaymentType(paymentType, ROUTES.CASH_REGISTER));
 
   return {
-    ...getPageDispatchers(dispatch, ownProps, 'Transaction', ROUTES.CASH_REGISTER),
+    ...getPageDispatchers(dispatch, 'Transaction', ROUTES.CASH_REGISTER),
     onUpdatePaymentType,
   };
 };
```
Comment (wlthomson) [inline src/pages/CashRegisterPage.js]: Can't do "suggested changes" on deleted lines it seems 🤷‍♂ 

```
const mapDispatchToProps = dispatch => {
   return {
    ...getPageDispatchers(dispatch, 'Transaction', ROUTES.CASH_REGISTER),
    onUpdatePaymentType: paymentType => dispatch(updatePaymentType(paymentType, ROUTES.CASH_REGISTER));
  }
}
```
